components/file_system/src/iosnoop/biosnoop.rs

```diff
@@ -146,52 +145,19 @@ pub fn get_io_type() -> IOType {
     unsafe { *IDX.with(|idx| IO_TYPE_ARRAY[idx.0]) }
 }
 
-unsafe fn get_io_stats() -> Option<HashMap<IOType, IOStats>> {
-    if let Some(ctx) = BPF_CONTEXT.as_mut() {
-        let mut map = HashMap::default();
-        for e in ctx.stats_table.iter() {
-            let io_type = ptr::read(e.key.as_ptr() as *const IOType);
-            let read = std::intrinsics::atomic_load(e.value.as_ptr() as *const u64);
-            let write = std::intrinsics::atomic_load((e.value.as_ptr() as *const u64).add(1));
-            map.insert(io_type, IOStats { read, write });
-        }
-        Some(map)
-    } else {
-        None
-    }
-}
-
-pub struct IOContext {
-    io_stats_map: Option<HashMap<IOType, IOStats>>,
-}
-
-impl IOContext {
-    pub fn new() -> Self {
-        IOContext {
-            io_stats_map: unsafe { get_io_stats() },
-        }
-    }
-
-    pub fn delta_and_refresh(&mut self) -> HashMap<IOType, IOStats> {
-        if self.io_stats_map.is_some() {
-            if let Some(map) = unsafe { get_io_stats() } {
-                for (io_type, stats) in &map {
-                    self.io_stats_map
-                        .as_mut()
-                        .unwrap()
-                        .entry(*io_type)
-                        .and_modify(|e| {
-                            e.read = stats.read - e.read;
-                            e.write = stats.write - e.write;
-                        })
-                        .or_insert(stats.clone());
-                }
-
-                return self.io_stats_map.replace(map).unwrap();
+pub(crate) fn fetch_io_bytes(io_type: IOType) -> IOBytes {
+    unsafe {
+        if let Some(ctx) = BPF_CONTEXT.as_mut() {
+            let mut buf = [0 as u8; std::mem::size_of::<IOType>()];
+            std::ptr::write(buf.as_mut_ptr() as *mut IOType, io_type);
+            if let Ok(e) = ctx.stats_table.get(&mut buf) {
+                let read = std::intrinsics::atomic_load(e.as_ptr() as *const u64) as i64;
+                let write = std::intrinsics::atomic_load((e.as_ptr() as *const u64).add(1)) as i64;
```

Comment:
btw what does `add(1)` mean here?

Reply:
Similar to my previous comment these casts to `*mut u64` don't look sound because they assume the `Vec<u8>` returned by `bcc::Table::get` is at least u64-aligned, and, unless `Table` is provide some special guarantee abut its return value, Rust does not guarantee the the underlying buffer here is u64-aligned.

Furthermore, I don't see why atomics are needed here at all. `bcc::Table::get` returns an owned `Vec<u8>` - there's no way another thread has access to this data.

These atomic loads look like they can be changed to safe conversions from `&[u8]` to u64. The easiest way I know of offhand to that conversion is with [byte_slice_cast::AsSliceOf](https://docs.rs/byte-slice-cast/1.0.0/byte_slice_cast/trait.AsSliceOf.html).

Reply:
Furthermore, this double-cast to a u64 pointer _then_ to an i64 seems suspicous. They are both 64-bits. Why not cast directly to an i64 pointer?

Reply:
Sorry for all the comments, but I just need to note that offhand I don't know the most simple and correct way to convert from `&[u8]` to `u64`. There must be some function somewhere that does this correctly in the presence of misalignment, but I don't know where to find it.

One part of the solution is [`u64::from_ne_bytes`](https://doc.rust-lang.org/stable/std/primitive.u64.html#method.from_ne_bytes) but to call that you first need a `[u8; 8]`.

To get that array you can probably use [`slice::as_chunks`](https://doc.rust-lang.org/stable/std/primitive.slice.html#method.as_chunks).

Reply:
@brson the same buffer is passed via eBPF and modified in kernel space, so they can be accessed concurrently.

Reply:
@brson you're right. I was thought that the `iter()` returns a ptr of the value, but actually, it's a ptr of on-stack vec.  So the atomics are not necessary here. 

For the u64-alignment, cause the definition of `stats_table` in `biosnoop.c` is 
```c
struct stats_t {
  u64 read;
  u64 write;
};
BPF_HASH(stats_by_type, io_type, struct stats_t);
```, seems we can sure that the returned value should be u64-alignment.

Reply:
I don't think the stats_t has any effect here. Maybe kernel side the memory is aligned, but the `rust-bcc` actually copied them into an on-stack vector ([code](https://github.com/rust-bpf/rust-bcc/blob/c4e5ef83568b220c8bf271a6f80e9579b2df1e5e/src/table.rs#L65)), making the alignment here suspicious.

Reply:
I got what you mean. Seems we can simply write like that `ptr::read_unaligned(e.value.as_ptr() as *const IOStats)`.

Reply:
I take a look at the example of rust-bcc and it doesn't consider the alignment either. https://github.com/rust-bpf/rust-bcc/blob/master/examples/contextswitch.rs#L123 @brson Is there anything promising it implicitly?

Reply:
it does look like the map value is obtained via `bpf_map_lookup_elem(fd, key, value)` call and the result is stored in an owned buffer from rust-bcc. I don't seems to find thread-safe guarantee of `bpf_map_lookup_elem` call. But anyway, let's remove the use of atomic here.